.\" Automatically generated by Pandoc 3.1.1
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "Error" "JS" "June 2, 2023" "JavaScript" "JavaScript Reference Manual"
.hy
.SH NAME
.PP
Error - Error
.SH SYNOPSIS
.PP
\f[B]\f[VB]Error\f[B]\f[R] objects are thrown when runtime errors occur.
The \f[V]Error\f[R] object can also be used as a base object for
user-defined exceptions.
See below for standard built-in error types.
.SH DESCRIPTION
.PP
Runtime errors result in new \f[V]Error\f[R] objects being created and
thrown.
.PP
\f[V]Error\f[R] is a serializable object, so it can be cloned with
\f[V]structuredClone()\f[R] or copied between Workers using
\f[V]postMessage()\f[R].
.SS Error types
.PP
Besides the generic \f[V]Error\f[R] constructor, there are other core
error constructors in JavaScript.
For client-side exceptions, see Exception handling statements.
.TP
\f[B]EvalError\f[R]
Creates an instance representing an error that occurs regarding the
global function \f[V]eval()\f[R].
.TP
\f[B]RangeError\f[R]
Creates an instance representing an error that occurs when a numeric
variable or parameter is outside its valid range.
.TP
\f[B]ReferenceError\f[R]
Creates an instance representing an error that occurs when
de-referencing an invalid reference.
.TP
\f[B]SyntaxError\f[R]
Creates an instance representing a syntax error.
.TP
\f[B]TypeError\f[R]
Creates an instance representing an error that occurs when a variable or
parameter is not of a valid type.
.TP
\f[B]URIError\f[R]
Creates an instance representing an error that occurs when
\f[V]encodeURI()\f[R] or \f[V]decodeURI()\f[R] are passed invalid
parameters.
.TP
\f[B]AggregateError\f[R]
Creates an instance representing several errors wrapped in a single
error when multiple errors need to be reported by an operation, for
example by \f[V]Promise.any()\f[R].
.TP
\f[B]InternalError\f[R] \f[I](non-standard)\f[R]
Creates an instance representing an error that occurs when an internal
error in the JavaScript engine is thrown.
E.g.
\[lq]too much recursion\[rq].
.SH CONSTRUCTOR
.TP
\f[B]Error()\f[R]
Creates a new \f[V]Error\f[R] object.
.SH STATIC METHODS
.TP
\f[B]Error.captureStackTrace()\f[R] \f[I](non-standard)\f[R]
A non-standard V8 function that creates the \f[V]stack\f[R] property on
an Error instance.
.TP
\f[B]Error.stackTraceLimit\f[R] \f[I](non-standard)\f[R]
A non-standard V8 numerical property that limits how many stack frames
to include in an error stacktrace.
.TP
\f[B]Error.prepareStackTrace()\f[R] \f[I](non-standard)\f[R] \f[I](optional)\f[R]
A non-standard V8 function that, if provided by usercode, is called by
the V8 JavaScript engine for thrown exceptions, allowing the user to
provide custom formatting for stacktraces.
.SH INSTANCE PROPERTIES
.PP
These properties are defined on \f[V]Error.prototype\f[R] and shared by
all \f[V]Error\f[R] instances.
.TP
\f[B]Error.prototype.constructor\f[R]
The constructor function that created the instance object.
For \f[V]Error\f[R] instances, the initial value is the \f[V]Error\f[R]
constructor.
.TP
\f[B]Error.prototype.name\f[R]
Represents the name for the type of error.
For \f[V]Error.prototype.name\f[R], the initial value is
\f[V]\[dq]Error\[dq]\f[R].
Subclasses like \f[V]TypeError\f[R] and \f[V]SyntaxError\f[R] provide
their own \f[V]name\f[R] properties.
.TP
\f[B]Error.prototype.stack\f[R] \f[I](non-standard)\f[R]
A non-standard property for a stack trace.
.PP
These properties are own properties of each \f[V]Error\f[R] instance.
.TP
\f[B]cause\f[R]
Error cause indicating the reason why the current error is thrown \[em]
usually another caught error.
For user-created \f[V]Error\f[R] objects, this is the value provided as
the \f[V]cause\f[R] property of the constructor\[cq]s second argument.
.TP
\f[B]columnNumber\f[R] \f[I](non-standard)\f[R]
A non-standard Mozilla property for the column number in the line that
raised this error.
.TP
\f[B]fileName\f[R] \f[I](non-standard)\f[R]
A non-standard Mozilla property for the path to the file that raised
this error.
.TP
\f[B]lineNumber\f[R] \f[I](non-standard)\f[R]
A non-standard Mozilla property for the line number in the file that
raised this error.
.TP
\f[B]message\f[R]
Error message.
For user-created \f[V]Error\f[R] objects, this is the string provided as
the constructor\[cq]s first argument.
.SH INSTANCE METHODS
.TP
\f[B]Error.prototype.toString()\f[R]
Returns a string representing the specified object.
Overrides the \f[V]Object.prototype.toString()\f[R] method.
.SH EXAMPLES
.SS Throwing a generic error
.PP
Usually you create an \f[V]Error\f[R] object with the intention of
raising it using the \f[V]throw\f[R] keyword.
You can handle the error using the \f[V]try...catch\f[R] construct:
.IP
.nf
\f[C]
try {
  throw new Error(\[dq]Whoops!\[dq]);
} catch (e) {
  console.error(\[ga]${e.name}: ${e.message}\[ga]);
}
\f[R]
.fi
.SS Handling a specific error type
.PP
You can choose to handle only specific error types by testing the error
type with the error\[cq]s \f[V]constructor\f[R] property or, if
you\[cq]re writing for modern JavaScript engines, \f[V]instanceof\f[R]
keyword:
.IP
.nf
\f[C]
try {
  foo.bar();
} catch (e) {
  if (e instanceof EvalError) {
    console.error(\[ga]${e.name}: ${e.message}\[ga]);
  } else if (e instanceof RangeError) {
    console.error(\[ga]${e.name}: ${e.message}\[ga]);
  }
  // etc.
  else {
    // If none of our cases matched leave the Error unhandled
    throw e;
  }
}
\f[R]
.fi
.SS Differentiate between similar errors
.PP
Sometimes a block of code can fail for reasons that require different
handling, but which throw very similar errors (i.e.\ with the same type
and message).
.PP
If you don\[cq]t have control over the original errors that are thrown,
one option is to catch them and throw new \f[V]Error\f[R] objects that
have more specific messages.
The original error should be passed to the new \f[V]Error\f[R] in the
constructor\[cq]s \f[V]options\f[R] parameter as its \f[V]cause\f[R]
property.
This ensures that the original error and stack trace are available to
higher-level try/catch blocks.
.PP
The example below shows this for two methods that would otherwise fail
with similar errors (\f[V]doFailSomeWay()\f[R] and
\f[V]doFailAnotherWay()\f[R]):
.IP
.nf
\f[C]
function doWork() {
  try {
    doFailSomeWay();
  } catch (err) {
    throw new Error(\[dq]Failed in some way\[dq], { cause: err });
  }
  try {
    doFailAnotherWay();
  } catch (err) {
    throw new Error(\[dq]Failed in another way\[dq], { cause: err });
  }
}

try {
  doWork();
} catch (err) {
  switch (err.message) {
    case \[dq]Failed in some way\[dq]:
      handleFailSomeWay(err.cause);
      break;
    case \[dq]Failed in another way\[dq]:
      handleFailAnotherWay(err.cause);
      break;
  }
}
\f[R]
.fi
.RS
.PP
\f[B]Note:\f[R] If you are making a library, you should prefer to use
error cause to discriminate between different errors emitted \[em]
rather than asking your consumers to parse the error message.
See the error cause page for an example.
.RE
.PP
Custom error types can also use the \f[V]cause\f[R] property, provided
the subclasses\[cq] constructor passes the \f[V]options\f[R] parameter
when calling \f[V]super()\f[R].
The \f[V]Error()\f[R] base class constructor will read
\f[V]options.cause\f[R] and define the \f[V]cause\f[R] property on the
new error instance.
.IP
.nf
\f[C]
class MyError extends Error {
  constructor(message, options) {
    // Need to pass \[ga]options\[ga] as the second parameter to install the \[dq]cause\[dq] property.
    super(message, options);
  }
}

console.log(new MyError(\[dq]test\[dq], { cause: new Error(\[dq]cause\[dq]) }).cause);
// Error: cause
\f[R]
.fi
.SS Custom error types
.PP
You might want to define your own error types deriving from
\f[V]Error\f[R] to be able to \f[V]throw new MyError()\f[R] and use
\f[V]instanceof MyError\f[R] to check the kind of error in the exception
handler.
This results in cleaner and more consistent error handling code.
.PP
See \[lq]What\[cq]s a good way to extend Error in
JavaScript?\[rq] (https://stackoverflow.com/questions/1382107/whats-a-good-way-to-extend-error-in-javascript)
on StackOverflow for an in-depth discussion.
.RS
.PP
\f[B]Warning:\f[R] Builtin subclassing cannot be reliably transpiled to
pre-ES6 code, because there\[cq]s no way to construct the base class
with a particular \f[V]new.target\f[R] without
\f[V]Reflect.construct()\f[R].
You need additional
configuration (https://github.com/loganfsmyth/babel-plugin-transform-builtin-extend)
or manually call
\f[V]Object.setPrototypeOf(this, CustomError.prototype)\f[R] at the end
of the constructor; otherwise, the constructed instance will not be a
\f[V]CustomError\f[R] instance.
See the TypeScript
FAQ (https://github.com/microsoft/TypeScript/wiki/FAQ#why-doesnt-extending-built-ins-like-error-array-and-map-work)
for more information.
.RE
.RS
.PP
\f[B]Note:\f[R] Some browsers include the \f[V]CustomError\f[R]
constructor in the stack trace when using ES2015 classes.
.RE
.IP
.nf
\f[C]
class CustomError extends Error {
  constructor(foo = \[dq]bar\[dq], ...params) {
    // Pass remaining arguments (including vendor specific ones) to parent constructor
    super(...params);

    // Maintains proper stack trace for where our error was thrown (only available on V8)
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, CustomError);
    }

    this.name = \[dq]CustomError\[dq];
    // Custom debugging information
    this.foo = foo;
    this.date = new Date();
  }
}

try {
  throw new CustomError(\[dq]baz\[dq], \[dq]bazMessage\[dq]);
} catch (e) {
  console.error(e.name); // CustomError
  console.error(e.foo); // baz
  console.error(e.message); // bazMessage
  console.error(e.stack); // stacktrace
}
\f[R]
.fi
.SH SEE ALSO
.IP \[bu] 2
A polyfill of
\f[V]Error\f[R] (https://github.com/zloirock/core-js#ecmascript-error)
with modern behavior like support \f[V]cause\f[R] is available in
\f[V]core-js\f[R] (https://github.com/zloirock/core-js)
.IP \[bu] 2
\f[V]throw\f[R]
.IP \[bu] 2
\f[V]try...catch\f[R]
.IP \[bu] 2
The V8 documentation (https://v8.dev/docs/stack-trace-api) for
\f[V]Error.captureStackTrace()\f[R], \f[V]Error.stackTraceLimit\f[R],
and \f[V]Error.prepareStackTrace()\f[R].
